### ISSUE TITLE
Bridge to Kubernetes GA

### ISSUE THREAD

**Issue by greenie-msft** [microsoft/mindaro#29]
We are super excited to announce General Availability of Bridge to Kubernetes.

Formerly known as Local Process with Kubernetes, Bridge to Kubernetes is an iterative development tool offered in Visual Studio and VS Code that allows developers to write, test and debug microservice code on their development workstations while consuming dependencies and inheriting existing configuration from a Kubernetes environment.

Developers using Bridge to Kubernetes can:

Increase efficiency by running code directly on development workstations.
Develop and debug microservice code end-to-end in isolation from teammates.
Separate development workflows from DevOps configuration.
Satisfy external microservice dependencies by connecting to existing services running in the cluster.
Learn how to use Bridge to Kubernetes by following our quickstarts for Visual Studio and VS Code.

**Comment by jmhill** [comment]
@greenie-msft  A few questions

1) Is there more information anywhere about the Linux support status/what scenarios are supported there? Tried the quickstart in both VS Code Remote WSL and VS Code Remote Container (devcontainer) and ran into problems when trying to use the new launch config (the task fails with "ClassName":"System.InvalidOperationException","Message":"Failed to run EndpointManager with elevated privileges: No such file or directory"). Should I open a ticket with this info for you guys or is this supposed to be unsupported right now?

2) I'm unfamiliar with this nip.io service that is being used now - my basic question is, will this work out of the box with an internal load balancer on AKS (we've now been told that not exposing our dev/pr review apps to the public internet is going to be a requirement for us)?

**Comment by igsys** [comment]
I was following the docs, and went up to this step and start having trouble.  The following is what the doc describes in the Debug Menu. 

[https://code.visualstudio.com/docs/containers/bridge-to-kubernetes](url)

![choose-bridge-to-kubernetes](https://user-images.githubusercontent.com/13977296/95534344-874af580-0a20-11eb-93e2-d975c0987681.png)

And this what I am actually getting.  I removed all the azure extensions from VSCode, start from a clean state, and follow each step in the documentation.

![CleanShot 2020-10-09 at 11 08 26](https://user-images.githubusercontent.com/13977296/95534303-671b3680-0a20-11eb-88c7-ae36e1ca2c43.png)

How to resolve this issue?

**Comment by pragyamehta** [comment]
Hi @igsys In the docs you have provided a link for, under this section [https://code.visualstudio.com/docs/containers/bridge-to-kubernetes#_connect-to-your-cluster-and-debug-a-service](Connect to your cluster and debug a service), there is a step 
`Open mindaro/samples/BikeSharingApp/Bikes from the Bike Sharing sample application in Visual Studio Code.`
Once you do that, you should see the correct launch profiles as displayed by the screenshot.

Please don't forget to follow the intermediate steps - selecting the correct namespace and running `npm install` before you start to debug.

Kindly open a new issue in case you have any more questions!

**Comment by daniv-msft** [comment]
@jmhill Thank you for contacting us.
1. As of today, we support full Linux machines, but not WSL/devcontainer. That said, these are on our radar (WSL especially) as other customers mentioned this already.
To help us prioritize this, could you please let us know if this is a blocker as of today for you? Is it possible for you to use a full Linux machine in the meantime (or Windows/Mac)?

2. Yes, I don't foresee any issues with nip.io in this case: nip.io is only operating at the DNS request level, and as such doesn't need to have public access to what is targeted.
For context, we need nip.io for load balancers when working isolated in Bridge. This is because when isolated, we provide an isolated subdomain to the endpoint, which is not possible to do with a load balancer's raw IP (for example, `http://jmhill.1.2.3.4/` isn't valid). Using nip.io allows us to bypass this.

**Comment by jmhill** [comment]
Hey @daniv-msft thanks so much for getting back to me: for question 1, it's a blocker for us in the sense that VS Code + .devcontainer is our preferred development environment now - our company uses windows machines (and I tested BTK on the windows side and it seems to work great) but all of our new projects embrace containerized development environments. BTK looks (like AZDS connect looked) to be the streamlined tool we'd want to use, but we favor the devcontainer usage enough that in the short term we'd probably stitch together other tools (maybe skaffold + telepresence or something similar) to achieve the "run project locally with proxy to cluster" approach if BTK + devcontainer isn't ready.

Thanks for the background on point 2, that helped alot.

**Comment by daniv-msft** [comment]
Thank you @jmhill for your reply!
Looking more into the devcontainer scenario, I don't see any fundamental blocker that would prevent us from running into this scenario. Looking at the error you received above, I suspect this is just an issue when we determine the path of our binaries. While we didn't spend enough time investigating this scenario, I would like to determine if I can unblock you and your team quickly.

Would it be possible for you to send me the Bridge to Kubernetes logs corresponding to this issue? You can find them here:
* For Windows: %TEMP%/Bridge to Kubernetes
* For OSX/Linux: $TMPDIR/Bridge to Kubernetes

Please include the latest `bridge-library-*`, `bridge-mindarocli-*` and `bridge-endpointmanager-*` (if created) after reproducing the bug.

If you don't feel comfortable uploading the logs on GitHub (as they might contain personally identifiable information), feel free to send them to me directly: `daniv@microsoft.com`.

**Comment by tvvignesh** [comment]
Hi. Great to see this. The sidecar support would be the most important thing in my point of view. I use the cluster with Linkerd service mesh which is going to inject a lot of sidecars. Supporting Bridge to Kubernetes with Sidecars would be awesome.

Tracked here: https://github.com/microsoft/mindaro/issues/52

Currently, I guess, I am going to have to disable all sidecars in dev cluster now.

**Comment by daniv-msft** [comment]
Thank you @tvvignesh for giving a try to Bridge and this feedback! Yes, the two items you mention are important for us, and are high in our backlog. I'll make sure we update the issues you linked when we have more update on these.

**Comment by tvvignesh** [comment]
@daniv-msft Hi. Thanks for your quick reply. I am really eager to try this but am actually facing more issues than actually mentioned here. I have documented it all here: https://github.com/microsoft/mindaro/issues/63

Hope its prioritized sometime. Thanks for your effort.

**Comment by daniv-msft** [comment]
@tvvignesh Thanks, we'll reply on the other issue you created to provide more info there. Thank you for this very complete status!

**Comment by jmhill** [comment]
@daniv-msft just closing the loop on our exchange above; I emailed the requested log files (copied out of the devcontainer that was failing) to the email address you provided. Thanks again for your help.

**Comment by daniv-msft** [comment]
Thank you @jmhill! We're investigating your logs and will provide more details on the email you sent us.

**Comment by alanmbarr** [comment]
Are there plans to release an updated VSIX for the Visual Studio IDE? Using any Kubernetes cluster is a nice feature on vscode so far.

**Comment by daniv-msft** [comment]
@alanmbarr We're still fixing a couple of issues/completing the implementation, but we're almost there. It should be available in Visual Studio in a couple of weeks now.
+@danegsta who drives this on our side.

**Comment by amir-davoodi** [comment]
Hi, I am trying to evaluate Bridge to Kubernetes in Visual Studio to use it within my team, however, the create profile dialogue fails to load the namespaces in my cluster! Is there anything other than connecting to the cluster needs to be done to get this working?

**Comment by daniv-msft** [comment]
Thanks @amir-davoodi for giving a try to Bridge to Kubernetes, and sorry you encounter this issue.
Could what you see be an occurrence of https://github.com/microsoft/mindaro/issues/26 ?

**Comment by toanstt** [comment]
![image](https://user-images.githubusercontent.com/1669489/99482337-408cdb80-298e-11eb-85fa-3cfb7b5ba42b.png)
I follow the document steps but stuck at running command "Bridge to Kubernetes: Configure"
Any idea to solve this one?

**Comment by amir-davoodi** [comment]
Thanks @daniv-msft for the reply. And yes! it looks like we are facing the same issue since we have enabled AAD for our AKS clusters. I'm looking forward for the fix. Is there any ETA?

**Comment by daniv-msft** [comment]
Thanks @amir-davoodi for your reply! Did you have a chance to try the workaround mentioned there? While we are working on making the experience better, in the foreseeable future we'll still rely on the kubectl authentication flow so that we don't need to reimplement it ourselves.
Do you encounter any issues while following [the instructions mentioned there](https://github.com/microsoft/mindaro/issues/26#issuecomment-696391176)?

**Comment by abvaden** [comment]
I have this same issue as well when using Bridge with WSL. A good alternative when working with WSL is to use [telepresence](https://www.telepresence.io/) even though it has limitations in WSL as well it has allowed me to continue my work with only minor modifications to my debugging setup.

I have also experimented with using Bridge and VisualStudio Online / Codespaces and encounter the same issue.

**Comment by liujiyuan** [comment]
Header propagation over http works fine.
How to propagation Header over tcp in envoy when Running in isolation mode?

**Comment by daniv-msft** [comment]
Thanks @liujiyuan for adding this comment. Tracking the replies to your question in https://github.com/microsoft/mindaro/issues/91

**Comment by williamohara** [comment]
@daniv-msft 
> 
> 1. As of today, we support full Linux machines, but not WSL/devcontainer. That said, these are on our radar (WSL especially) as other customers mentioned this already.
>    To help us prioritize this, could you please let us know if this is a blocker as of today for you? Is it possible for you to use a full Linux machine in the meantime (or Windows/Mac)?

This is an aggravation for me - not so much a blocker - yet. I need to have parallelism. I will have to do a lot of context switching between WSL and Windows - when do you think you will support WSL?

**Comment by daniv-msft** [comment]
@williamohara Thanks for your message. We have made good progress on this since the previous comments, and we're close from having WSL support. Unless we discover major issues with the final integration, we should be able to provide this in the coming weeks (let's say end of February/beginning of March) in VS Code.

**Comment by williamohara** [comment]
@daniv-msft Well - is WSL coming ? I am waiting with bated breath. Let me know if you need another  beta customer to try it out. I can't imagine my use case would be comprehensive though. its just me contributing to an open source project. :)

**Comment by rakeshvanga** [comment]
@williamohara Bridge To Kubernetes now supports WSL. Please find the documentation [Kubernetes service environment variables](https://code.visualstudio.com/docs/containers/kubernetes-env-vars) on using the WSL with Bridge To Kubernetes.
Let me know if you face any issues.
Thanks!

**Comment by dariuszparys** [comment]
I see that there is work in progress for sidecar support https://github.com/microsoft/mindaro/projects/1#card-51374140 is there any estimation when this can be expected or progress of the work? (Branch, etc...)?

**Comment by Brice187** [comment]
> Support in remote ssh would be very important in my case since I do all my development in a remote VM. Mentioned here: [#14 (comment)](https://github.com/microsoft/mindaro/issues/14#issuecomment-697594883)

Please add remote ssh for **Bridge to Kubernetes**. This would be awesome!
